uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/TOILRequestTest.php

```diff
@@ -676,4 +652,107 @@ public function testDeleteForAbsenceTypeWithinAGivenPeriodWhenNonExpiredOnlyFlag
+      'toil_to_accrue' => 2,
+      'duration' => 120,
+      'expiry_date' => CRM_Utils_Date::processDate('+100 days')
+    ]);
+
+    $toilRequest2 = TOILRequestFabricator::fabricateWithoutValidation([
+      'type_id' => $absenceType,
+      'contact_id' => 1,
+      'status_id' => 1,
+      'from_date' => CRM_Utils_Date::processDate('2016-04-10'),
+      'to_date' => CRM_Utils_Date::processDate('2016-04-10'),
+      'toil_to_accrue' => 2,
+      'duration' => 120,
+      'expiry_date' => CRM_Utils_Date::processDate('2016-12-10')
+    ]);
+
+    //Only the first toil request is valid and not expired, so will be deleted
+    $startDate = new DateTime('2016-03-10');
+    TOILRequest::deleteAllForAbsenceType($absenceType, $startDate, null, true);
+
+    //the expired balance change for the other TOIL will not be deleted
+    $balanceChanges = new LeaveBalanceChange();
+    $balanceChanges->find();
+    $this->assertEquals($balanceChanges->N, 1);
+    $balanceChanges->fetch();
+
+    $this->assertEquals($balanceChanges->source_id, $toilRequest2->id);
+
+    //the leave request for the other TOIL linked to expired balance changes will not be deleted
+    $leaveRequest = new LeaveRequest();
+    $leaveRequest->find();
+    $this->assertEquals($leaveRequest->N, 1);
+    $leaveRequest->fetch();
+
+    $this->assertEquals($leaveRequest->id, $toilRequest2->leave_request_id);
+
+    $leaveRequestDate = new LeaveRequestDate();
+    $leaveRequestDate->find();
+    $this->assertEquals($leaveRequestDate->N, 1);
+    $leaveRequestDate->fetch();
+
+    //the leave request date for the other TOIL linked to expired balance changes will not be deleted
+    $this->assertEquals($leaveRequestDate->date, '2016-04-10');
+
+    $toilRequest = new ToilRequest();
+    $toilRequest->find();
+    $this->assertEquals($toilRequest->N, 1);
+    $toilRequest->fetch();
+    $this->assertEquals($toilRequest->id, $toilRequest2->id);
+  }
+
+  public function testDeleteForAbsenceTypeWithNoEndDateWhenNonExpiredOnlyFlagIsFalse() {
+    $absenceType = 1;
+    TOILRequestFabricator::fabricateWithoutValidation([
+      'type_id' => $absenceType,
+      'contact_id' => 1,
+      'status_id' => 1,
+      'from_date' => CRM_Utils_Date::processDate('2016-03-10'),
+      'to_date' => CRM_Utils_Date::processDate('2016-03-13'),
+      'toil_to_accrue' => 2,
+      'duration' => 120,
+      'expiry_date' => CRM_Utils_Date::processDate('+100 days')
+    ]);
+
+    TOILRequestFabricator::fabricateWithoutValidation([
+      'type_id' => $absenceType,
+      'contact_id' => 1,
+      'status_id' => 1,
+      'from_date' => CRM_Utils_Date::processDate('2016-04-10'),
+      'to_date' => CRM_Utils_Date::processDate('2016-04-10'),
+      'toil_to_accrue' => 2,
+      'duration' => 120,
+      'expiry_date' => CRM_Utils_Date::processDate('2016-12-10')
+    ]);
+
+    $startDate = new DateTime('2016-03-10');
+    TOILRequest::deleteAllForAbsenceType($absenceType, $startDate, null, false);
```

Comment:
This comment applies to all the "deleteSomething" method tests:

If you don't assert that the records we want to delete actually exist before the deletion, there's no way to know you actually deleted something. For example, if you simply remove the 2 calls to the TOILRequestFabricator, this test will still pass, even though nothing was actually deleted.